hnix-store-core/src/System/Nix/Nar.hs

```diff
@@ -0,0 +1,216 @@
+{-# LANGUAGE KindSignatures      #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TupleSections       #-}
+{-# LANGUAGE TypeApplications    #-}
+
+{-|
+Description : Allowed effects for interacting with Nar files.
+Maintainer  : Shea Levy <shea@shealevy.com>
+|-}
+module System.Nix.Nar (
+    FileSystemObject(..)
+  , IsExecutable (..)
+  , Nar(..)
+  , getNar
+  , localUnpackNar
+  , putNar
+  ) where
+
+import           Control.Applicative
+import           Control.Monad              (replicateM, replicateM_)
+import Control.Monad.Writer
+import qualified Data.Binary                as B
+import qualified Data.Binary.Get            as B
+import qualified Data.Binary.Put            as B
+import qualified Data.ByteString.Lazy.Char8 as BSL
+import           Data.Foldable              (forM_)
+import qualified Data.Map                   as Map
+import           Data.Maybe                 (fromMaybe)
+import           Data.Monoid                ((<>))
+import qualified Data.Text                  as T
+import qualified Data.Text.Encoding         as E
+import           GHC.Int                    (Int64)
+import           System.Directory
+import           System.FilePath
+import           System.Posix.Files         (createSymbolicLink)
+
+import           System.Nix.Path
+
+
+data NarEffects (m :: * -> *) = NarEffets {
+    readFile :: FilePath -> m BSL.ByteString
+  , listDir  :: FilePath -> m [FileSystemObject]
+}
+
+
+-- Directly taken from Eelco thesis
+-- https://nixos.org/%7Eeelco/pubs/phd-thesis.pdf
+
+data Nar = Nar { narFile :: FileSystemObject }
+    deriving (Eq, Show)
+
+data FileSystemObject =
+    Regular IsExecutable BSL.ByteString
+  | Directory (Map.Map PathName FileSystemObject)
```

Comment:
`PathName` is incorrect here. `PathName` is limited to just what Nix allows in the *top-level* name of an output, but files within that can be named anything. The real rule here is an arbitrary byte-string with no nulls or `/`s.